[Pipeliner] Implement dynamic loop peeling

sjw36 opened the pull request:
* wrap epilogue stages with `if` statements in case ub <= lb + num_stages
- enabled for tritonamdgpu-stream-pipeline

For instance, when pipelining with num_stages==3

```
load (0)
load(1)
local_store(0)
for (0..K-1) {
  dot(i)
  load(i+2)
  local_store(i+1)
}
if (K > 2) { // iteration K-2
  dot(K-2)
  local_store(K-1)
}
if (K > 1) { // iteration K-1
  dot(K-1)
}
```

ThomasRaoux reviewed (commented):
please upstream the expander changes along with unit tests in MLIR first. Also I haven't looked in details but I don't think the peeling code looks like what we want as I would not expect block of operations within IF regions

antiagainst commented:
Yup; I've already let @sjw36 know about the expander's connection with upstream and requirements to update in tandem there. I believe @sjw36 will do that. This is just mirroring here to use the CI to run tests easier.

sjw36 commented:
> please upstream the expander changes along with unit tests in MLIR first. Also I haven't looked in details but I don't think the peeling code looks like what we want as I would not expect block of operations within IF regions

Thanks @ThomasRaoux. This is mostly to start the discussion and get your and @pawelszczerbuk feedback to the implementation. What do you recommend for guarding the peeled code. I guess we could create `select`s for all live-outs but that would still create an assignment which is not always valid.

sjw36 commented:
> > Thanks @ThomasRaoux. This is mostly to start the discussion and get your and @pawelszczerbuk feedback to the implementation. What do you recommend for guarding the peeled code. I guess we could create `select`s for all live-outs but that would still create an assignment which is not always valid.
> 
> I would expect the epilogue peeling to be unchanged with regards to stitching, we should only apply predication to ops like we currently do in the prologue/kernel

Currently `predicateFn` just deals with global memory loads to add masking. Since `tt.dot` operations from shared memory, the data would be naturally masked in the prologue, the dot could be executed unconditionally. But there are other ops that should not be executed such as atomics or global stores.

Would you suggest an enhancement to the predicateFn to allow unconditional tt.dot (or any compute), but only guard any global store related ops?

ThomasRaoux commented:
> > > Thanks @ThomasRaoux. This is mostly to start the discussion and get your and @pawelszczerbuk feedback to the implementation. What do you recommend for guarding the peeled code. I guess we could create `select`s for all live-outs but that would still create an assignment which is not always valid.
> > 
> > 
> > I would expect the epilogue peeling to be unchanged with regards to stitching, we should only apply predication to ops like we currently do in the prologue/kernel
> 
> Currently `predicateFn` just deals with global memory loads to add masking. Since `tt.dot` operations from shared memory, the data would be naturally masked in the prologue, the dot could be executed unconditionally. But there are other ops that should not be executed such as atomics or global stores.
> 
> Would you suggest an enhancement to the predicateFn to allow unconditional tt.dot (or any compute), but only guard any global store related ops?

predicateFn currently speculatively execute ops that don't have side effects and handle predication for others. Either we need to be able to predicate any side effecting op or we need to limit what we allowed to be in the epilogue

manman-ren commented:
I actually need this for flash attention + computation on NVGPU. But I don't quite get why we need to pass crossStageValues to emitEpilogue, I thought emitEpilogue will just use the valueMapping. Is the idea to use an if statement for each epilogue stage or use predicate for each op in the epilogue stage?

sjw36 commented:
> I actually need this for flash attention + computation on NVGPU. But I don't quite get why we need to pass crossStageValues to emitEpilogue, I thought emitEpilogue will just use the valueMapping. Is the idea to use an if statement for each epilogue stage or use predicate for each op in the epilogue stage?

We are moving to use predicate for each op, and select for intermediate results from epilogue stage. I'll post an update shortly.

sjw36 commented:
I've updated most of the changes.
* use predicateFn for epilogue ops
* conditionally select previous results vs stage outputs
* updated tt::predicateFn for tt.store

I'll update the tests next, and add more.

sjw36 commented:
Upstream PR here: https://github.com/llvm/llvm-project/pull/106436

sjw36 commented:
I removed AMD specifics to match upstream as much as possible. I'll follow up in a new PR with the AMD changes.

sjw36 commented:
@antiagainst please move out of draft.

antiagainst commented:
Thanks @sjw36! Can you also update the pull request summary accordingly? 

@pawelszczerbuk and @ThomasRaoux: I believe the changes to `lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp` are just copying https://github.com/llvm/llvm-project/pull/106436. So the only thing need review is the changes to lib/Dialect/TritonGPU/Transforms/Pipeliner/PipeliningUtility.cpp.

sjw36 commented:
> @pawelszczerbuk and @ThomasRaoux: I believe the changes to `lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp` are just copying [llvm/llvm-project#106436](https://github.com/llvm/llvm-project/pull/106436). So the only thing need review is the changes to lib/Dialect/TritonGPU/Transforms/Pipeliner/PipeliningUtility.cpp.

I removed these changes since dynamic pipeline peeling is not enabled for the default flow.

antiagainst commented:
Okay now this is just a strict copy of the upstream change. Let me go ahead and land.
